extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -0,0 +1,2040 @@
+                         .processorFactory(
+                             new SegmentGeneratorFrameProcessorFactory(
+                                 dataSchema,
+                                 columnMappings,
+                                 tuningConfig
+                             )
+                         )
+      );
+
+      return builder.build();
+    } else if (querySpec.getDestination() instanceof TaskReportMSQDestination) {
+      return queryDef;
+    } else {
+      throw new ISE("Unsupported destination [%s]", querySpec.getDestination());
+    }
+  }
+
+  private static DataSchema generateDataSchema(
+      MSQSpec querySpec,
+      RowSignature querySignature,
+      ClusterBy queryClusterBy,
+      ColumnMappings columnMappings
+  )
+  {
+    final DataSourceMSQDestination destination = (DataSourceMSQDestination) querySpec.getDestination();
+    final boolean isRollupQ = isRollupQuery(querySpec.getQuery());
+
+    final Pair<List<DimensionSchema>, List<AggregatorFactory>> dimensionsAndAggregators =
+        makeDimensionsAndAggregatorsForIngestion(
+            querySignature,
+            queryClusterBy,
+            destination.getSegmentSortOrder(),
+            columnMappings,
+            isRollupQ,
+            querySpec.getQuery()
+        );
+
+    return new DataSchema(
+        destination.getDataSource(),
+        new TimestampSpec(ColumnHolder.TIME_COLUMN_NAME, "millis", null),
+        new DimensionsSpec(dimensionsAndAggregators.lhs),
+        dimensionsAndAggregators.rhs.toArray(new AggregatorFactory[0]),
+        makeGranularitySpecForIngestion(isRollupQ, querySpec.getQuery()),
+        new TransformSpec(null, Collections.emptyList())
+    );
+  }
+
+  private static GranularitySpec makeGranularitySpecForIngestion(boolean isRollupQ, Query<?> query)
+  {
+    if (!isRollupQ) {
+      return new ArbitraryGranularitySpec(Granularities.NONE, false, Intervals.ONLY_ETERNITY);
+    } else {
+      final String queryGranularity = query.getContextValue(GroupByQuery.CTX_TIMESTAMP_RESULT_FIELD_GRANULARITY, "");
+
+      if (checkIfTimeColumnsAreEqual((GroupByQuery) query) && !queryGranularity.isEmpty()) {
+        return new ArbitraryGranularitySpec(
+            Granularity.fromString(queryGranularity),
+            true,
+            Intervals.ONLY_ETERNITY
+        );
+      }
+      return new ArbitraryGranularitySpec(Granularities.NONE, true, Intervals.ONLY_ETERNITY);
+    }
+  }
+
+
+  /**
+   * Checks if the time columns present in the groupByQuery context are same. One is set by
+   * {@link DruidQuery#toGroupByQuery()} and the other is set by
+   * {@link MSQTaskQueryMaker#runQuery(DruidQuery)}
+   *
+   * @return true if both groupByQuery context values are present and equal else returns false.
+   */
+  private static boolean checkIfTimeColumnsAreEqual(GroupByQuery groupByQuery)
+  {
+    final String msqTimeColumn = groupByQuery.getContextValue(QueryKitUtils.CTX_TIME_COLUMN_NAME, "");
```

Comment:
This lets the query determine the time column name? Is this a general feature? I've seen other places where we hard-code the assumption that the time column is named `__time`. Can each datasource have its own time column name? Each segment? (I.e. each ingest?)

Is this something we should track in the proposed catalog?